nemo/collections/asr/models/k2_sequence_models.py

```diff
@@ -0,0 +1,393 @@
+    ):
+        """
+        Forward pass of the model.
+
+        Args:
+            input_signal: Tensor that represents a batch of raw audio signals,
+                of shape [B, T]. T here represents timesteps, with 1 second of audio represented as
+                `self.sample_rate` number of floating point values.
+            input_signal_length: Vector of length B, that contains the individual lengths of the audio
+                sequences.
+            processed_signal: Tensor that represents a batch of processed audio signals,
+                of shape (B, D, T) that has undergone processing via some DALI preprocessor.
+            processed_signal_length: Vector of length B, that contains the individual lengths of the
+                processed audio sequences.
+
+        Returns:
+            A tuple of 3 elements -
+            1) The log probabilities tensor of shape [B, T, D].
+            2) The lengths of the acoustic sequence after propagation through the encoder, of shape [B].
+            3) The greedy token predictions of the model of shape [B, T] (via argmax)
+        """
+        # trying to load token_lm from token_lm_cache_dict or token_lm_path if it hasn't been loaded yet
+        if self.use_graph_lm and self.token_lm is None:
+            if self.token_lm_cache_dict is not None:
+                logging.info(f"""Loading token_lm from the dict cache at the first .forward() call.""")
+                self.token_lm = k2.Fsa.from_dict(self.token_lm_cache_dict)
+                self.token_lm_cache_dict = None
+            elif self.token_lm_path is not None:
+                logging.warning(f"""Loading token_lm from `{self.token_lm_path}` at the first .forward() call.""")
+                self.token_lm = load_graph(self.token_lm_path)
+                if self.token_lm is None:
+                    raise ValueError(f"""Failed to load token_lm""")
+            else:
+                raise ValueError(f"""Failed to load token_lm""")
+            self.loss.update_graph(self.token_lm)
+            if self.use_graph_lm:
+                self.transcribe_decoder.update_graph(self.token_lm)
+
+        log_probs, encoded_len, greedy_predictions = super().forward(
+            input_signal=input_signal,
+            input_signal_length=input_signal_length,
+            processed_signal=processed_signal,
+            processed_signal_length=processed_signal_length,
+        )
+
+        # greedy_predictions from .forward() are incorrect for criterion_type=`map`
+        # getting correct greedy_predictions, if needed
+        if self.use_graph_lm and (not self.training or self.transcribe_training):
+            greedy_predictions, encoded_len, _ = self.transcribe_decoder.forward(
+                log_probs=log_probs, log_probs_length=encoded_len
+            )
+        return log_probs, encoded_len, greedy_predictions
+
+
+class EncDecK2SeqModelBPE(EncDecCTCModelBPE):
```

Comment:
Since the mixin commit [3ed4f2b](https://github.com/NVIDIA/NeMo/pull/3351/commits/3ed4f2b9c4dfd9359d92b0771ef010c1bab380c7), this file became quite small, and I don't think it'll be modified often in future.